packages/agreement-readmodel-writer-sql/src/readModelService.ts

```diff
@@ -0,0 +1,202 @@
+import {
+  Agreement,
+  AgreementDocument,
+  AgreementDocumentId,
+  AgreementId,
+} from "pagopa-interop-models";
+import {
+  AgreementReadModelService,
+  agreementDocumentToAgreementDocumentSQL,
+} from "pagopa-interop-readmodel";
+import { and, eq, lte } from "drizzle-orm";
+import {
+  agreementAttributeInReadmodelAgreement,
+  agreementConsumerDocumentInReadmodelAgreement,
+  agreementContractInReadmodelAgreement,
+  agreementInReadmodelAgreement,
+  agreementStampInReadmodelAgreement,
+  DrizzleReturnType,
+  DrizzleTransactionType,
+} from "pagopa-interop-readmodel-models";
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export function readModelServiceBuilder(
+  db: DrizzleReturnType,
+  agreementReadModelService: AgreementReadModelService
+) {
+  const updateMetadataVersionInAgreementRelatedTables = async (
+    tx: DrizzleTransactionType,
+    agreementId: AgreementId,
+    newVersion: number
+  ): Promise<void> => {
+    const agreementRelatedTables = [
+      agreementStampInReadmodelAgreement,
+      agreementAttributeInReadmodelAgreement,
+      agreementConsumerDocumentInReadmodelAgreement,
+      agreementContractInReadmodelAgreement,
+    ];
+
+    await tx
+      .update(agreementInReadmodelAgreement)
+      .set({ metadataVersion: newVersion })
+      .where(
+        and(
+          eq(agreementInReadmodelAgreement.id, agreementId),
+          lte(agreementInReadmodelAgreement.metadataVersion, newVersion)
+        )
+      );
+
+    for (const table of agreementRelatedTables) {
+      await tx
+        .update(table)
+        .set({ metadataVersion: newVersion })
+        .where(
+          and(
+            eq(table.agreementId, agreementId),
+            lte(table.metadataVersion, newVersion)
+          )
+        );
+    }
+  };
+
+  return {
+    async upsertAgreement(
+      agreement: Agreement,
+      metadataVersion: number
+    ): Promise<void> {
+      return await agreementReadModelService.upsertAgreement(
+        agreement,
+        metadataVersion
+      );
+    },
+    async deleteAgreement(
+      agreementId: AgreementId,
+      metadataVersion: number
+    ): Promise<void> {
+      return await agreementReadModelService.deleteAgreementById(
+        agreementId,
+        metadataVersion
+      );
+    },
+    async upsertConsumerDocument(
```

Comment:
Why didn't you put all these subsequent functions inside the common `packages/readmodel/src/agreementReadModelService.ts` and instead define them here? We could centralize all the agreement readmodel queries in the share agreementReadModelService and just use it as it is in all the services that need it

Reply:
See discussion below https://github.com/pagopa/interop-be-monorepo/pull/1631#discussion_r2014316548